src/UnscentedTransformKalmanInversion.jl

```diff
@@ -0,0 +1,474 @@
+        X = construct_successful_perturbation(uki, u_p, u_p_mean, successful_ens)
+        Y = construct_successful_perturbation(uki, g_ext, g_mean_ext, successful_ens)
+        # Create/Enlarge buffers if needed
+        tmp = get_buffer(get_process(uki)) # the buffer stores Y' * Γ_inv of [size(Y,2),size(Y,1)]
+        ys1, ys2 = size(Y)
+        if length(tmp) == 0  # no buffer
+            push!(tmp, zeros(ys2, ys1)) # stores Y' * Γ_inv
+            push!(tmp, zeros(ys2, ys2)) # stores Y' * Γ_inv * Y
+        elseif (size(tmp[1], 1) < ys2) || (size(tmp[1], 2) < ys1) # existing buffer is too small
+            tmp[1] = zeros(ys2, ys1)
+            tmp[2] = zeros(ys2, ys2)
+        end
+
+        if process.impose_prior
+            lmul_obs_noise_cov_inv!(view(tmp[1]', 1:size(g, 1), 1:ys2), uki, Y[1:size(g, 1), :], g_idx) # store in transpose, with view helping reduce allocations
+            view(tmp[1]', (size(g, 1) + 1):ys1, 1:ys2) .= process.Σ_ν_scale * prior_cov_inv * Y[(size(g, 1) + 1):end, :]
+        else
+            lmul_obs_noise_cov_inv!(view(tmp[1]', :, 1:ys2), uki, Y, g_idx) # store in transpose, with view helping reduce allocations
+        end
+        view(tmp[1], 1:ys2, :) .*= inv_noise_scaling
+
+        ### Check internal multiplication
+        #=
+        Σ_ν_inv =  get_obs_noise_cov_inv(uki)[g_idx, g_idx] * get_Δt(uki)[end] / process.Σ_ν_scale
+                multmat = Y' * Σ_ν_inv
+                @info "diff" norm(multmat - tmp[1][1:ys2,1:ys1])
+        =#
+        ###
+
+        tmp[2][1:ys2, 1:ys2] = tmp[1][1:ys2, 1:ys1] * Y
+
+        for i in 1:ys2
+            tmp[2][i, i] += 1.0
+        end
+
+        Ω = inv(tmp[2][1:ys2, 1:ys2]) # Ω = (I + Y' * Γ_inv * Y)^-1 = I - Y' (Y Y' + Γ_inv)^-1 Y      
+        u_mean = u_p_mean + X * FT.(Ω * tmp[1][1:ys2, 1:ys1] * (y_ext .- g_mean_ext)) #  mean update = Ω * Y' * Γ_inv * (y .- g_mean))
+        uu_cov = X * Ω * X' # cov update
+
+        ### Check agains true Kalman gain:
+        #=
+        Gain = X*Y'*(inv(Y*Y'+get_obs_noise_cov(uki) ./ inv_noise_scaling))
+        gain_mean = (u_p_mean + Gain * (y_ext .- g_mean_ext))
+        @info norm(gain_mean - u_mean)
+        =#
+        ###
+
+        ########### Save results
+        process.obs_pred[end][g_idx] .= g_mean
+        process.u_mean[end][u_idx] .= u_mean
+        process.uu_cov[end][u_idx, u_idx] .= uu_cov
+
+    end
+    function failsafe_update(uki, u, g, u_idx, g_idx, failed_ens)
+        #perform analysis on the model runs
+        succ_gauss_analysis!(uki, u, g, u_idx, g_idx, failed_ens)
+        #perform new prediction output to model parameters u_p
+        u_p = update_ensemble_prediction!(process, get_Δt(uki)[end], u_idx)
+        return u_p
+    end
+    return FailureHandler{TransformUnscented, SampleSuccGauss}(failsafe_update)
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+UKI analysis step  : g is the predicted observations  `Ny x N_ens` matrix
+"""
+function update_ensemble_analysis!(
```

Comment:
Maybe I'm misunderstanding, but is it necessary to have both `update_ensemble_analysis!` and `succ_gauss_analysis!`? Seems to be a lot of code duplication.

Reply:
I think this is something for a future PR. I have been unhappy with the way we have implemented the failure handler. but basically this is the reason for this. It is much broader than this method though, so I think it will have to persist for now.

Reply:
Created an issue.